apisix/healthcheck_manager.lua

```diff
@@ -0,0 +1,322 @@
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local require = require
+local ipairs   = ipairs
+local pcall   = pcall
+local exiting      = ngx.worker.exiting
+local pairs    = pairs
+local tostring = tostring
+local core = require("apisix.core")
+local config_local   = require("apisix.core.config_local")
+local healthcheck
+local events = require("apisix.events")
+local tab_clone = core.table.clone
+local timer_every = ngx.timer.every
+local string_sub     = string.sub
+
+local _M = {}
+local working_pool = {}     -- resource_path -> {version = ver, checker = checker}
+local waiting_pool = {}      -- resource_path -> resource_ver
+
+local DELAYED_CLEAR_TIMEOUT = 10
+local healthcheck_shdict_name = "upstream-healthcheck"
+local is_http = ngx.config.subsystem == "http"
+if not is_http then
+    healthcheck_shdict_name = healthcheck_shdict_name .. "-" .. ngx.config.subsystem
+end
+
+
+local function get_healthchecker_name(value)
+    return "upstream#" .. (value.resource_key or value.upstream.resource_key)
+end
+_M.get_healthchecker_name = get_healthchecker_name
+
+
+local function remove_etcd_prefix(key)
+    local prefix = ""
+    local local_conf = config_local.local_conf()
+    local role = core.table.try_read_attr(local_conf, "deployment", "role")
+    local provider = core.table.try_read_attr(local_conf, "deployment", "role_" ..
+    role, "config_provider")
+    if provider == "etcd" and local_conf.etcd and local_conf.etcd.prefix then
+        prefix = local_conf.etcd.prefix
+    end
+    return string_sub(key, #prefix + 1)
+end
+
+
+local function fetch_latest_conf(resource_path)
+    local resource_type, id
+    -- Handle both formats:
+    -- 1. /<etcd-prefix>/<resource_type>/<id>
+    -- 2. /<resource_type>/<id>
+    resource_path = remove_etcd_prefix(resource_path)
+    resource_type, id = resource_path:match("^/([^/]+)/([^/]+)$")
+    if not resource_type or not id then
+        core.log.error("invalid resource path: ", resource_path)
+        return nil
+    end
+
+    local key
+    if resource_type == "upstreams" then
+        key = "/upstreams"
+    elseif resource_type == "routes" then
+        key = "/routes"
+    elseif resource_type == "services" then
+        key = "/services"
+    elseif resource_type == "stream_routes" then
+        key = "/stream_routes"
+    else
+        core.log.error("unsupported resource type: ", resource_type)
+        return nil
+    end
+
+    local data = core.config.fetch_created_obj(key)
+    if not data then
+        core.log.error("failed to fetch configuration for type: ", key)
+        return nil
+    end
+    local resource = data:get(id)
+    if not resource then
+        -- this can happen if the resource was deleted
+        -- after the this function was called so we don't throw error
+        core.log.warn("resource not found: ", id, " in ", key)
```

Comment:
print more information, help the OPS to confirm the reason

`this can happen if the resource was deleted`